torchrl/modules/functional_modules.py

```diff
@@ -0,0 +1,227 @@
+from copy import deepcopy
+
+import torch
+from torch import nn
+
+from torchrl.data import TensorDict
+from torchrl.data.tensordict.tensordict import TensorDictBase
```

Comment:
Some quick questions to better understand TensorDict:
- Are all tensors required to be on the same device? What about other properties, like dtype?
- Does TensorDict optimize storage? Will it try to pack the tensors together in memory?

Reply:
Device: yes as of now it does require all tensors to be on the same device. I think I see where you're going with this though: models can be split across devices. What we want to support is tensordict.to(device) but we could also have unset devices:
TensorDict({"a": torch.randn(3, device=0)}, []) # unset

TensorDict({"a": torch.randn(3, device=0)}, [], device=0) # set

In the second case all tensors must be on the same device, in the first they can be placed on different ones.

Dtype: no we don't require that by design. We want to store things like masks, booleans, integers (eg discrete actions) etc.  We also want to close the door to algebraic operations as I don't see the value of doing multiplications over tensors of different dtypes, so to say.
If you want to map all tensors to a dtype, you can do
TensorDict.apply_(lambda x: x.to(dtype))

Storage: since the dtype and shape are different across tensors I don't really see how we could optimise storage but I'm happy to hear your thoughts!